### PR TITLE
CancelProofEnvoyStream

### PR DESCRIPTION
**Pull request by carloseltuerto**
Description: Wrapper for the EnvoyStream guaranteeing that a another stream method can't be called after a "cancel"
Risk Level: None - Cronvoy only
Testing: CI
Docs Changes: N/A
Release Notes: N/A
Signed-off-by: Charles Le Borgne <cleborgne@google.com>

### CODE REVIEW

File: test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    mSendDataBlock.open();
+    mReadDataBlock.open();
+    // Wait for all the Threads to die.
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  private class MockedStream extends EnvoyHTTPStream {
```
Comment (RyanTheOptimist) [inline test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java:347]: Can you add a comment which explains the various Block/Latch behaviors of this class?

File: test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void cancel_withoutStreamSet() {
+    // Does nothing because the "Concurrently Running Stream Operations Count" starts with "1".
+    // This is the desired outcome - the cancel is postponed. Once the stream is set, the next
+    // Stream Operation will invoke "cancel".
+    cancelProofEnvoyStream.cancel();
```
Comment (RyanTheOptimist) [inline test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java:254]: Should this check that the cancel invocation count is 0?

**Review by RyanTheOptimist — DISMISSED** [review]
Looks great. Two tiny comments.

File: test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java (hunk 2 above)
Comment (carloseltuerto) [reply test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java:254]: It is the `MockedStream` that can increment the cancel invocation count, and this test does not invoke `setStream`: it is pointless to verify it.

Added a comment.

File: test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java (hunk 1 above)
Comment (carloseltuerto) [reply test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java:347]: Added a comment. Short story: a `ConditionVariable` is in reality a `CountDownLatch` where the initial count is "1".

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**